core/src/main/java/org/apache/calcite/rel/rules/DphypJoinReorderRule.java

```
@@ -0,0 +1,86 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.rel.rules;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.plan.RelRule;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rex.RexBuilder;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.tools.RelBuilder;
+
+import org.immutables.value.Value;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/** Rule that re-orders a {@link Join} tree using dphyp algorithm.
+ *
+ * @see CoreRules#HYPER_GRAPH_OPTIMIZE */
+@Value.Enclosing
+public class DphypJoinReorderRule
+    extends RelRule<DphypJoinReorderRule.Config>
+    implements TransformationRule {
+
+  protected DphypJoinReorderRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    HyperGraph hyperGraph = call.rel(0);
+    RelBuilder relBuilder = call.builder();
+    // make all field name unique and convert the
+    // HyperEdge condition from RexInputRef to RexInputFieldName
+    hyperGraph.convertHyperEdgeCond();
+
+    // enumerate by Dphyp
+    DpHyp dpHyp = new DpHyp(hyperGraph, relBuilder, call.getMetadataQuery());
+    dpHyp.startEnumerateJoin();
+    RelNode orderedJoin = dpHyp.getBestPlan();
+    if (orderedJoin == null) {
+      return;
```

If we get here, it means something went wrong, doesn't it?
IIANM this would lead to a CannotPlanException (if we cannot decompose the HyperGraph into Joins, the we'll keep the HyperGraph which is a "not implementable" operator). In that case, maybe it would be a good idea to throw an exception (to fail early and with a clearer message) instead of simply returning (and failing later with a more generic and "cryptic" exception)

---

My original intention is that when the query is too complex, even dphyp can be very time-consuming. So, if the execution of `dpHyp.startEnumerateJoin` exceeds a certain time limit, the enumeration process should be forcibly terminated. In this case, the dp table may not have results yet. After returning null, the hypergraph simplification algorithm might be executed.
Alternatively, the rules related to dphyp could be integrated into a single method, with the original Join tree as the input parameter. If the optimization times out, the original Join tree is returned, and other Join reorder algorithms can be called subsequently.
Do you have any suggestions for this? I will update it accordingly.